framework/core/instance.cpp

```diff
@@ -381,20 +381,54 @@ void Instance::query_gpus()
 
 PhysicalDevice &Instance::get_suitable_gpu()
 {
-	assert(!gpus.empty() && "No physical devices were found on the system.");
+	if (gpus.size() == 0)
+	{
+		throw std::runtime_error("No physical devices found.");
+	}
+
+	uint32_t count_device_type[VK_PHYSICAL_DEVICE_TYPE_CPU + 1];
+	memset(count_device_type, 0, sizeof(count_device_type));
+
+	for (uint32_t i = 0; i < gpus.size(); i++)
+	{
+		const auto deviceType = gpus[i]->get_properties().deviceType;
+		assert(deviceType <= VK_PHYSICAL_DEVICE_TYPE_CPU);
+		count_device_type[deviceType]++;
+	}
+
+	const VkPhysicalDeviceType device_type_preference[] = {
+	    VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU,
+	    VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU,
+	    VK_PHYSICAL_DEVICE_TYPE_VIRTUAL_GPU,
+	    VK_PHYSICAL_DEVICE_TYPE_CPU,
+	    VK_PHYSICAL_DEVICE_TYPE_OTHER};
 
-	// Find a discrete GPU
-	for (auto &gpu : gpus)
+	VkPhysicalDeviceType search_for_device_type = VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU;
+	for (uint32_t i = 0; i < sizeof(device_type_preference) / sizeof(VkPhysicalDeviceType); i++)
 	{
-		if (gpu->get_properties().deviceType == VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU)
+		if (count_device_type[device_type_preference[i]])
 		{
-			return *gpu;
+			search_for_device_type = device_type_preference[i];
+			break;
 		}
 	}
 
```

Comment:
An integrated or a virtual GPU would still be better than a software implementation.

Reply:
And? This is exactly the point of my patch.